### PR TITLE
download: Add fallback to private GitHub repo

### PR DESCRIPTION
**Pull request by juliaogris**
If download of GitHub release fails with HTTP status 404 and the
`HERMIT_GITHUB_TOKEN` environment variable isn't empty attempt to treat
the URL like that of a private GitHub repo:

* Authenticate HTTP requests with Auth header
* Get `assetID` via Github API
* Download via `assetID`

### CODE REVIEW

File: cache/cache.go [hunk 1]
```diff
@@ -289,3 +303,93 @@ func (p2 *progressWriterAt) WriteAt(p []byte, off int64) (n int, err error) {
 	p2.task.Add(n)
 	return n, errors.WithStack(err)
 }
+
+// matches: https://github.com/{OWNER}/{REPO}/releases/download/{TAG}/{ARTEFACT}
```
Comment (camh-) [inline cache/cache.go]: s/{ARTEFACT}/{ASSET}/

File: cache/cache.go [hunk 2]
```diff
@@ -289,3 +303,93 @@ func (p2 *progressWriterAt) WriteAt(p []byte, off int64) (n int, err error) {
...
+type githubReleaseInfo struct {
+	owner, repo, tag, asset, token string
+}
+
+func getGithubReleaseInfo(uri string) (*githubReleaseInfo, bool) {
+	m := githubRe.FindStringSubmatch(uri)
+	if len(m) != 5 {
+		return nil, false
+	}
+	token := os.Getenv("HERMIT_GITHUB_TOKEN")
```
Comment (camh-) [inline cache/cache.go]: Move this before the regex check. no point doing that work if you are just going to bail out here anyway.

Perhaps also, use the same form as the others:

```
g := &githubReleaseInfo{}
g.token = os.Getenv(...)
```

File: cache/cache.go [hunk 3]
```diff
@@ -289,3 +303,93 @@ func (p2 *progressWriterAt) WriteAt(p []byte, off int64) (n int, err error) {
...
+		return nil, nil, errors.Wrap(err, "cannot GET private github repo assets via API")
+	}
+	defer resp.Body.Close()
+	r := []GHRelease{}
+	if err := json.NewDecoder(resp.Body).Decode(&r); err != nil {
+		return nil, nil, errors.Wrap(err, "cannot decode private github repo assets")
+	}
+	assetID, err := getAssetID(r, ghi.tag, ghi.asset)
+	header.Set("Accept", "application/octet-stream")
+	u = fmt.Sprintf("https://api.github.com/repos/%s/%s/releases/assets/%d", ghi.owner, ghi.repo, assetID)
```
Comment (camh-) [inline cache/cache.go]: I dont think you should construct this URL. Github gives it to you in the asset as the `url` field.

File: cache/cache.go [hunk 4]
```diff
@@ -215,10 +218,21 @@ func (c *Cache) downloadHTTP(b *ui.Task, checksum string, uri string, cachePath
...
+	w, response, err := Download(c.httpClient, nil, uri, downloadPath)
 	if err != nil {
 		return "", "", errors.WithStack(err)
 	}
+	// Check for potential private github release
+	if response.StatusCode == 404 {
+		if ghi, ok := getGithubReleaseInfo(uri); ok {
+			response.Body.Close()
+			w.Close()
+			w, response, err = downloadGHPrivate(c.httpClient, ghi, downloadPath) //nolint:bodyclose // it _is_ handled below.
```
Comment (camh-) [inline cache/cache.go]: the other nolint style in this file (and also in this PR) is `// nolint:` - i.e. a space after the comment marker. We can argue that that is wrong, but it is consistent. Either way, this or the other in this PR should be fixed (line 363).

File: cache/cache.go [hunk 5]
```diff
@@ -289,3 +303,93 @@ func (p2 *progressWriterAt) WriteAt(p []byte, off int64) (n int, err error) {
...
+		if r.TagName != tag {
+			continue
+		}
+		for _, a := range r.Assets {
+			if a.Name == assetName {
+				return a.ID, nil
+			}
+		}
+	}
+	return -1, fmt.Errorf("cannot find asset %s %s", tag, assetName)
```
Comment (camh-) [inline cache/cache.go]: I think you should return the `int` zero value (0) here for the error case. It is meaningless in the error case but setting it to anything other than the zero value implies it has meaning.

File: cache/cache.go (hunk 3 above)
Comment (juliaogris) [reply cache/cache.go]: oh!!

Comment (juliaogris) [reply cache/cache.go]: done, thank you!
kinda obvious in hindsight.

File: cache/cache.go (hunk 5 above)
Comment (juliaogris) [reply cache/cache.go]: changed to string and url.

**Review by camh- — APPROVED** [review]
:frog: LGTM :frog:

**Pull request closed without merging**